Source/Concurrency/CivlTypeChecker.cs

```diff
@@ -342,15 +359,32 @@ private void TypeCheckActionDecls()
       }
     }
 
-    private void TypeCheckActionImpls()
+    private void FinishActionInitialization()
     {
-      ActionVisitor actionVisitor = new ActionVisitor(this);
-      foreach (var action in Enumerable.Concat<Action>(AllAtomicActions, procToIntroductionAction.Values))
+      SkipAtomicAction.FinishInitialization();
+      var allActions = Enumerable.Concat<Action>(AllAtomicActions, procToIntroductionAction.Values);
+      foreach (var action in allActions)
       {
-        actionVisitor.VisitAction(action);
+        var impl = action.impl;
+        impl.OriginalBlocks = impl.Blocks;
+        impl.OriginalLocVars = impl.LocVars;
+      }
+      foreach (var action in allActions)
+      {
+        Inliner.ProcessImplementation(Options, program, action.impl);
+      }
+      foreach (var action in allActions)
+      {
+        var impl = action.impl;
+        impl.OriginalBlocks = null;
+        impl.OriginalLocVars = null;
+      }
```

Comment:
Maybe there should be a helper method for inlining (in `Inliner`?).

Reply:
I will do this in a separate PR also.